waspc/src/Wasp/Analyzer/TypeChecker/TypeError.hs

```diff
@@ -117,3 +97,38 @@ getWeakenErrorMessageAndCtx = getTypeCoercionErrorMessageAndCtx $
       [ concatPrefixAndText "Expected type: " (show t),
         concatPrefixAndText "Actual type:   " (show $ exprType texpr)
       ]
+
+getTypeCoercionErrorMessageAndCtx :: (Type -> TypedExpr -> String) -> TypeCoercionError -> (String, Ctx)
+getTypeCoercionErrorMessageAndCtx getUncoercableTypesMsg typeCoercionError = (fullErrorMsg, ctx)
+  where
+    (typesErrorMsg, ctxMsgs, ctx) = getUncoercableTypesMsgAndCtxMsgsAndCtx getUncoercableTypesMsg typeCoercionError
+    fullErrorMsg
+      | null ctxMsgs = typesErrorMsg
+      | otherwise = intercalate "\n\n" [typesErrorMsg, concatContextualMsgs ctxMsgs]
+
+-- | Takes a function for constructing a type coercion error message, and a `TypeCoercionError`.
+--
+-- It recursively traverses the error stack and returns a tuple containing:
+-- - The original type coercion error message
+-- - An array of contextual messages further explaining the original error
+-- - The error's context
+getUncoercableTypesMsgAndCtxMsgsAndCtx :: (Type -> TypedExpr -> String) -> TypeCoercionError -> (String, [String], Ctx)
```

Comment:
Since you went with a nice long name, I would add couple more words into it: `getUncoercableTypesErrorMsgAndErrorCtxMsgsAndParsingCtx`
because otherwise it is really hard to say what is Msg vs CtxMsgs vs Ctx -> it all sounds the same. But ErrorMsg vs ErrorCtxMsgs vs ParsingCtx -> well, still not great but maybe a bit better?

Ok, we have comment at least.

Reply:
Hm, I guess I thought I'd introduce the naming pattern:

- An **error message** is the string that describes the error.
- **Context messages** are strings describing the location (i.e., the context). They aren't necessarily error messages, as they can describe the context of anything (for example, a warning)
- **The Context** is the location of "the thing" described by the context message (of course, this name/concept already existed).

But that was pretty arbitrary and I didn't document it anywhere :smile: 

Perhaps a better name for _context messages_ is _context info_, I could then name the function one of the following:
- `getUncoercableTypesErrorMsgAndCtxInfoAndCtx`
- `getUncoercableTypesErrorMsgAndCtxInfoAndParsingCtx` - I'm not sure about this one since we're referring to the same thing with two different names. `Ctx` in `CtxInfo` and `ParsingCtx` are talking about the same `Ctx`.
- `getUncoercableTypesErrorMsgAndParsingCtxInfoAndParsingCtx`

Perhaps an even better solution is grouping the parsing context and its info messages into a same type, but this can probably wait for a different PR. I'd be happy to do it now if you think it makes sense, and if we can afford to spend a bit more time on this.

@Martinsos  What do you think?

Reply:
The thing is, it is not the same Ctx -> One thing is Parser.Ctx, while another thing are these error context messages.

I would propose following naming:
1. Error message -> string that describes the core part of error.
2. Error context messages (or Error context info) -> additional messages that provide more info about the error, mostly about its surroundings (context) in the Wasp lang AST. These messages are aware that we are dealing with an error, in a sense.
3. Parsing context (Parser.Ctx) -> information about where this piece of code is coming from, regarding the source file, where it was parsed from. Will be also important for LSP later. This information doesn't know anything about errors as a concept and is present in all our intermediate representations.

The reason I would not group (2) with (3) is that (3) is more general and present through all IRs, while (2) is relatively specific for this phase/IR, maybe one more (Evaluator). At least for now -> maybe more generalized version of it will emerge with time, but I don't think we are there yet.